kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -18,83 +18,257 @@
 
 import java.io.IOException;
 import java.util.List;
-import java.util.stream.Stream;
+import java.util.NoSuchElementException;
+import java.util.Optional;
 
 import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.*;
 import io.delta.kernel.fs.FileStatus;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.types.StructType;
 import io.delta.kernel.utils.CloseableIterator;
 import io.delta.kernel.utils.Tuple2;
 
-import io.delta.kernel.internal.actions.Action;
-import io.delta.kernel.internal.actions.AddFile;
-import io.delta.kernel.internal.actions.Metadata;
-import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.actions.*;
 import io.delta.kernel.internal.fs.Path;
-import io.delta.kernel.internal.lang.CloseableIterable;
 import io.delta.kernel.internal.lang.Lazy;
 import io.delta.kernel.internal.snapshot.LogSegment;
+import io.delta.kernel.internal.util.Logging;
+
+/**
+ * Replays a history of actions, resolving them to produce the current state of the table. The
+ * protocol for resolution is as follows:
+ *  - The most recent {@link AddFile} and accompanying metadata for any `(path, dv id)` tuple wins.
+ *  - {@link RemoveFile} deletes a corresponding AddFile. A RemoveFile "corresponds" to the AddFile
+ *    that matches both the parquet file URI *and* the deletion vector's URI (if any).
+ *  - The most recent {@link Metadata} wins.
+ *  - The most recent {@link Protocol} version wins.
+ *  - For each `(path, dv id)` tuple, this class should always output only one {@link FileAction}
+ *    (either AddFile or RemoveFile)
+ *
+ * This class exposes only two public APIs
+ * - {@link #loadProtocolAndMetadata()}: replay the log in reverse and return the first non-null
+ *                                       Protocol and Metadata
+ * - {@link #getAddFilesAsColumnarBatches}: return all active (not tombstoned) AddFiles as
+ *                                          {@link ColumnarBatch}s
+ */
+public class LogReplay implements Logging {
+
+    /** Read schema when searching for the latest Protocol and Metadata. */
+    public static final StructType PROTOCOL_METADATA_READ_SCHEMA = new StructType()
+        .add("protocol", Protocol.READ_SCHEMA)
+        .add("metaData", Metadata.READ_SCHEMA);
+
+    /**
+     * Read schema when searching for all the active AddFiles (need some RemoveFile info, too).
+     * Note that we don't need to read the entire RemoveFile, only the path and dv info.
+     */
+    public static final StructType ADD_REMOVE_READ_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA)
```

Comment:
furthermore, do not read the add file stats, since we won't be parsing them for a while (a few weeks)